edgedb/retry.py

```diff
@@ -0,0 +1,87 @@
+import asyncio
+import random
+
+from . import errors
+from . import transaction as _transaction
+
+
+DEFAULT_MAX_ITERATIONS = 3
+
+
+def default_backoff(attempt):
+    return (2 ** attempt) * 0.1 + random.randrange(100) * 0.1
+
+
+class AsyncIOIteration(_transaction.AsyncIOTransaction):
+    def __init__(self, retry, owner):
+        super().__init__(owner)
+        self.__retry = retry
+
+    async def start(self):
```

Comment:
Does this have to be a "public" method?

Reply:
This overrides the method on `Transaction`, so yes, it at least has to cover public method on the superclass.

Reply:
Maybe we rename `Transaction` to `BaseTransaction`, make all methods private on it, and then we can add a new `Transaction` class with public proxies for those methods. Then this class can extend `BaseTransaction` and wouldn't need to have a `public start` method, which is useless.

Reply:
I think it's nice that object that user gets is a `Transaction` object, not something special. We can document it. User can do `isinstance(db, Transaction)` and it works the same in `retry` and `try_transaction`, etc.

We could also emit a different object in `__aenter__` for managed transactions which doesn't have `start()` method, but we chose not to dive too deep. I think this is very similar case.

Reply:
> We could also emit a different object in __aenter__ for managed transactions which doesn't have start() method, but we chose not to dive too deep.

What I designed that API I don't think I put too much thought into what should be returned from `__aenter__`, tbh. So the decision wasn't super deliberate.

> I think it's nice that object that user gets is a Transaction object, not something special. We can document it. User can do isinstance(db, Transaction) and it works the same in retry and try_transaction, etc.

Maybe it would indeed be nice, but what's the point of calling `start()` manually inside `async with`?  Currently that seems to be allowed?

Reply:
No. You either do:
```
async with tx:
  something
```
Or you do:
```
transaction.start()
transaction.commit()
```
In the former case `_managed` is set to `True`. In the latter case it keeps `False`. So for the transaction created by `retry`, latter case throws `InterfaceError` on unmanaged transaction `start()`.

Reply:
> ... So for the transaction created by retry, latter case throws InterfaceError on unmanaged transaction start().

So 

```python
async for tx in con.retry():
  async with tx:
    await tx.start()
```

Will throw an InterfaceError? That's what I want. I don't think I've seen a test for that -- if there isn't one can you add it please?

Reply:
In your case it's "transaction is already started". But I've added few tests.

Reply:
Great, thanks